src/array/cuda/labor_sampling.cu

```diff
@@ -0,0 +1,858 @@
+
+  IdType out_row = blockIdx.x * TILE_SIZE + threadIdx.y;
+  const auto last_row = min(static_cast<IdType>(blockIdx.x + 1) * TILE_SIZE, num_rows);
+
+  constexpr FloatType ONE = 1;
+
+  while (out_row < last_row) {
+    const auto row = rows[out_row];
+
+    const auto in_row_start = indptr[row];
+    const auto out_row_start = subindptr[out_row];
+
+    const IdType d = indptr[row + 1] - in_row_start;
+
+    if (d > 0) {
+      const auto k = min(num_picks, d);
+      // slightly better than NS
+      const FloatType d_ = ds ? ds[row] : d;
+      FloatType var_target = d_ * d_ / k + (ds ? d2s[row] - d_ * d_ / d : 0);
+
+      auto c = cs[out_row];
+      const int num_valid = min(d, (IdType)CTA_SIZE);
+      FloatType var_1;
+      do {
+        var_1 = 0;
+        if (A) {
+          for (int idx = threadIdx.x; idx < d; idx += CTA_SIZE) {
+            const auto w = A[in_row_start + idx];
+            const auto ps = probs ? probs[out_row_start + idx] : w;
+            var_1 += w * w / min(ONE, c * ps);
+          }
+        } else {
+          for (int idx = threadIdx.x; idx < d; idx += CTA_SIZE) {
+            const auto ps = probs[out_row_start + idx];
+            var_1 += 1 / min(ONE, c * ps);
+          }
+        }
+        var_1 = BlockReduce(temp_storage).Sum(var_1, num_valid);
+        if (threadIdx.x == 0)
+          var_1_bcast[threadIdx.y] = var_1;
+        __syncthreads();
+        var_1 = var_1_bcast[threadIdx.y];
+
+        c *= var_1 / var_target;
+      } while (min(var_1, var_target) / max(var_1, var_target) < 1 - eps);
+
+      if (threadIdx.x == 0)
+        cs[out_row] = c;
+    }
+
+    out_row += BLOCK_CTAS;
+  }
+}
+
+template<typename IdType, int BLOCK_SIZE, IdType TILE_SIZE>
+__global__ void map_vertex_ids_global(
+  const IdType * const global,
+  IdType * const new_global,
+  const IdType num_vertices,
+  const runtime::cuda::DeviceOrderedHashTable<IdType> table) {
+  assert(BLOCK_SIZE == blockDim.x);
+
+  using Mapping = typename OrderedHashTable<IdType>::Mapping;
+
+  const IdType tile_start = TILE_SIZE * blockIdx.x;
+  const IdType tile_end = min(TILE_SIZE * (blockIdx.x + 1), num_vertices);
+
+  for (IdType idx = threadIdx.x + tile_start; idx < tile_end; idx += BLOCK_SIZE) {
+    const Mapping& mapping = *table.Search(global[idx]);
+    new_global[idx] = mapping.local;
+  }
+}
+
+}  // namespace
+
+/////////////////////////////// CSR ///////////////////////////////
+
+template <DGLDeviceType XPU, typename IdType, typename FloatType>
+std::pair<COOMatrix, FloatArray> CSRLaborSampling(CSRMatrix mat,
+                  IdArray NIDs,
+                  IdArray rows_arr,
+                  const int64_t num_picks__,
+                  FloatArray prob_arr,
+                  IdArray random_seed,
+                  IdArray cnt,
+                  int importance_sampling) {
+  constexpr bool use_csr_spmv = false;
+  constexpr bool use_hashtable = false;
+  const IdType num_picks = num_picks__;
```

Comment:
Most other variables come in arrays of IdType, when I want to use std::min for example between a number of IdType and num_picks, then I would need an extra cast. This avoids those types of inconveniencies.

Reply:
num_picks can be at most as much as number of nodes in the graph. That is why it makes sense that it is of type IdType, otherwise I would need to keep casting it and treat it in a special way. CPU side was using it only once so I changed it but CUDA side seems to use it a lot, so I do the casting once and for all.

Reply:
OK.  I would prefer the argument name without the double underscores though, since they are more user-facing.